sklearn/linear_model/_glm/tests/test_glm.py

```diff
@@ -29,6 +64,474 @@ def regression_data():
+    X = X[:, :-1]  # remove intercept
+    X = np.concatenate((X, X), axis=0)
+    assert np.linalg.matrix_rank(X) <= min(n_samples, n_features)
+    y = np.r_[y, y]
+    if fit_intercept:
+        coef = coef_with_intercept
+        intercept = coef[-1]
+        coef = coef[:-1]
+    else:
+        coef = coef_without_intercept
+        intercept = 0
+    model.fit(X, y)
+
+    rtol = 3e-5
+    assert model.intercept_ == pytest.approx(intercept, rel=rtol)
+    assert_allclose(model.coef_, coef, rtol=rtol)
+
+
+@pytest.mark.parametrize("solver", SOLVERS)
+@pytest.mark.parametrize("fit_intercept", [True, False])
+def test_glm_regression_unpenalized(solver, fit_intercept, glm_dataset):
+    """Test that unpenalized GLM converges for all solvers to correct solution.
+
+    We work with a simple constructed data set with known solution.
+    Note: This checks the minimum norm solution for wide X, i.e.
+    n_samples < n_features:
+        min ||w||_2 subject to w minimizing the mean deviviance.
+    """
+    model, X, y, coef, _, _ = glm_dataset
+    n_samples, n_features = X.shape
+    alpha = 0  # unpenalized
+    params = dict(
+        alpha=alpha,
+        fit_intercept=fit_intercept,
+        # solver=solver,  # only lbfgs available
+        tol=1e-12,
+        max_iter=1000,
+    )
+
+    model = clone(model).set_params(**params)
+    if fit_intercept:
+        X = X[:, :-1]  # remove intercept
+        intercept = coef[-1]
+        coef = coef[:-1]
+    else:
+        intercept = 0
+
+    if n_samples < n_features:
+        warnings.filterwarnings("ignore", category=scipy.linalg.LinAlgWarning)
+        warnings.filterwarnings("ignore", category=ConvergenceWarning)
+    model.fit(X, y)
+
+    # FIXME: `assert_allclose(model.coef_, coef)` should work for all cases but fails
+    # for the wide/fat case with n_features > n_samples. Most current GLM solvers do
+    # NOT return the minimum norm solution with fit_intercept=True.
+    rtol = 5e-5
+    if n_samples > n_features:
+        assert model.intercept_ == pytest.approx(intercept)
+        assert_allclose(model.coef_, coef, rtol=rtol)
+    else:
+        # As it is an underdetermined problem, prediction = y. The following shows that
+        # we get a solution, i.e. a (non-unique) minimum of the objective function ...
+        assert_allclose(model.predict(X), y, rtol=1e-6)
+        assert_allclose(
+            model._get_loss().link.inverse(X @ coef + intercept), y, rtol=5e-7
+        )
+        if fit_intercept:
+            # But it is not the minimum norm solution. Otherwise the norms would be
+            # equal.
+            norm_solution = (1 + 1e-12) * np.linalg.norm(np.r_[intercept, coef])
+            norm_model = np.linalg.norm(np.r_[model.intercept_, model.coef_])
+            assert norm_model > norm_solution
+            pytest.xfail(
+                reason=f"GLM with {solver} does not provide the minimum norm solution."
+            )
```

Comment:
I think this is not really an valid use of `xfail`: as far as I know, lbfgs has no implicit inductive bias to select the minimum normal solution (but I believe SGD has it).

So we could instead remove the xfail statement here and instead check that re-running the same fit with a very small value of `alpha` (e.g. `alpha = 1e-30`) would actually lead to converging close the minimum norm solution of the unregularized problem.

WDYT?

Reply:
> We could, however, add it on top as additional check.

Yes this is what I meant:

- first check that lbfgs with alpha=0 converges to a non-minimal normal solution (same predictions and loss as the minimum norm solution to the the constructed unpenalized problem of the fixture)
- then refit lbfgs with alpha=eps and checks that it converges to the minimal norm solution of the constructed unpenalized problem.

Reply:
I tried it and it does not work out: Setting tiny alpha with lbfgs does not give the minimum norm solution. I don't know if that is due to the solver or due to GLMs. This is kind of uncharted territory where I do not know any literature.

See https://github.com/scikit-learn/scikit-learn/pull/23619/commits/382d17707457638ee936bf44a9f25bb3e296a71c.

Reply:
This is mysterious but I guess making this kind of assertions is too dependent on details such as the precise stopping criteria of LBFGS, the condition number of the hessian near in the subspace of solutions for the unpenalized problem.

Thanks for trying.